onecodex/models/collection.py

```diff
@@ -389,6 +391,171 @@ def taxonomy(self):
+                    else:
+                        raise OneCodexException(
+                            f"Functional profile for sample {sample_id} not successful."
+                        )
+                else:
+                    functional_profiles.append(functional_profile[0])
+            else:
+                raise OneCodexException(
+                    f"More than one ({len(functional_profile)}) functional analyses found for sample {sample_id}"
+                )
+        # ensure all the functional profiles are the same job version
+        job_ids = set([obj.job.id for obj in functional_profiles])
+        if len(job_ids) > 1:
+            raise OneCodexException(
+                "FunctionalProfiles contain multiple analysis job versions: {}".format(
+                    ", ".join(job_ids)
+                )
+            )
+
+        self._cached["functional_profiles"] = functional_profiles
+
+    @property
+    def _functional_profiles(self):
+        if "functional_profiles" not in self._cached:
+            self._functional_profiles_fetch()
+
+        return self._cached["functional_profiles"]
+
+    def _collate_functional_results(
+        self, annotation, metric, taxa_stratified=True, fill_missing=False, filler=0
+    ):
+        """
+        Return a dataframe of all functional profile data.
+
+        Parameters
+        ----------
+        annotation : {'pathways', 'metacyc', 'eggnog', 'go', 'ko', 'ec', 'pfam', 'reaction'}, optional
+            Annotation data to return
+        taxa_stratified : 'bool', optional
+            Return taxonomically stratified data
+        metric : {'coverage', 'abundance'} for annotation=='pathways', {'rpk', 'cpm'} for other annotations
+            Metric values to return
+        fill_missing : 'bool', optional
+            Fill np.nan values
+        filler : 'float', optional
+            Value with which to fill np.nans
+        """
+        # validate args
+        valid_annotation_values = [
+            "pathways",
+            "metacyc",
+            "eggnog",
+            "go",
+            "ko",
+            "ec",
+            "pfam",
+            "reaction",
+        ]
+        if annotation not in valid_annotation_values:
+            raise ValueError(f"'annotation' must be one of {', '.join(valid_annotation_values)}")
+        if annotation == "pathways":
+            if metric not in ["coverage", "abundance"]:
+                raise ValueError(
+                    "if using annotation='pathways', 'value' must be one of ['coverage', 'abundance']"
+                )
+        elif metric not in ["cpm", "rpk"]:
+            raise ValueError(
+                f"if using annotation={annotation}, 'value' must be one of ['cpm', 'rpk']"
+            )
+
+        data = {}
+        all_features = set()
+        # iterate over functional profiles, subset data, and store in data dict
+        for profile in self._functional_profiles:
+            # get table from mainline
```

Comment:
nit: since this is a public repo, we try not to refer (too much) to internal component names like mainline. it'd be better to say something like "get table using One Codex API"

Reply:
done